enhancements/microshift/audit-log-configuration-options.md

```diff
@@ -0,0 +1,249 @@
+| WriteRequestBodies          | In addition to logging metadata for all requests, logs request bodies for every write request to the API servers (create, update, patch, delete, deletecollection). This profile has more resource overhead than the Default profile. |
+| AllRequestBodies            | In addition to logging metadata for all requests, logs request bodies for every read and write request to the API servers (get, list, create, update, patch). This profile has the most resource overhead.                            |
+
+Both OpenShift and kube-apiserver allow defining the depth of collected API request (metadata, request, response).  OpenShift extends the definition of these profiles to filter messages based on the request's verb (see table above). It will therefore be necessary to implement logic in MicroShift to support these extended definitions. This is possible and can be done via the Policy.audit.k8s.io/v1 object. Each kube-apiserver level can be extended to filter per verb, per API resource.  Thus, MicroShift will have to know how to map a given profile to a backend Policy object such that the message scope, verb and resource exactly represent the OpenShift profile.
+
+The Kubernetes Policy spec can be found [here](https://kubernetes.io/docs/reference/config-api/apiserver-audit.v1/#audit-k8s-io-v1-Policy).
+
+A simple example of the OpenShift "Default" profile could be represented in a Policy object as:
+
+```yaml
+apiVersion: audit.k8s.io/v1
+kind: Policy
+metadata:
+  name: Default
+omitStages:
+- "RequestReceived"
+rules:
+- level: Metadata
+    verbs:
+    - create
+    - update
+    - patch
+    - delete
+    - deletecollection
+    - get
+    - list
+    - create
+    - update
+    - patch
+    resources:
+    - group: "*"
+      resources: "*"
+```
+
+Note that "omitStages" is determines the point at which a message is logged. For more on stages, see [here](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/). Omitting the "RequestRecieved" stage is preferable to cut down on log spam but may be useful for debugging. 
+
+### Risks and Mitigations
+
+- Exceeding disk capacity: Microshift targets small form-factor devices with limit on-board storage. If the product of `maxFileSize` and `maxFiles` equals a size larger than the available storage, the apiserver risks destabilizing the system by maxing out the consumed storage.  This can be mitigated via documentation which recommends the user understand their storage limitations when setting these values. Logging a warning during startup when the apiserver values exceed the available storage at the logging path. It is impossible to predict which component of MicroShift or the host system will fail when the disk is out of space, thus logging a calculated overage can provide an important clue.
```

Comment:
Logging at startup is great, but it would not address the problem in many cases when disks are filled up over many days of runtime. Should we also create a watchdog for this to run every X minutes and log a message if the condition is encountered?

Reply:
That starts to veer back towards "Actions on Reaching Disk Capacity," which I believe we extracted to a separate ticket.

Reply:
How do users monitor storage capacity for other applications? Can we suggest they use the same approach, instead of adding more complexity to MicroShift?

Reply:
That data is exposed in OpenShift via it's [monitoring stack](https://docs.openshift.com/container-platform/4.14/monitoring/monitoring-overview.html), but that's not deployed in MicroShift.  Aside from that, cockpit would be the only other way I'm aware of.

Exposing metrics is a whole feature unto itself though and not in scope for this EP.

Reply:
I'm not sure if it is useful to log this information on MicroShift startup then.

Reply:
Perhaps then we take special care to call this out in documentation, and do not perform the check described here. Wdyt?

Reply:
I've gone ahead and removed the portion about logging a potential overage.  Documentation should be sufficient.